tokens/cnft-vault/programs/cnft-vault/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+            AccountMeta::new_readonly(ctx.accounts.tree_authority.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.leaf_owner.key(), true),
+            AccountMeta::new_readonly(ctx.accounts.leaf_delegate.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.new_leaf_owner.key(), false),
+            AccountMeta::new(ctx.accounts.merkle_tree.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.log_wrapper.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.compression_program.key(), false),
+            AccountMeta::new_readonly(ctx.accounts.system_program.key(), false),
+        ];
+
+        // first 8 bytes of SHA256("global:transfer")   
+        let transfer_discriminator: [u8;8] = [163, 52, 200, 231, 140, 3, 69, 186];//hex::decode("a334c8e78c0345ba").expect("hex decode fail"); 
+        //msg!("{:?}", transfer_discriminator);
+        
+        let mut data: Vec<u8> = vec![];
+        data.extend(transfer_discriminator);
+        data.extend(root);
+        data.extend(data_hash);
+        data.extend(creator_hash);
+        data.extend(nonce.to_le_bytes());
+        data.extend(index.to_le_bytes());
+
+        let mut account_infos: Vec<AccountInfo> = vec![
+            ctx.accounts.tree_authority.to_account_info(),
+            ctx.accounts.leaf_owner.to_account_info(),
+            ctx.accounts.leaf_delegate.to_account_info(),
+            ctx.accounts.new_leaf_owner.to_account_info(),
+            ctx.accounts.merkle_tree.to_account_info(),
+            ctx.accounts.log_wrapper.to_account_info(),
+            ctx.accounts.compression_program.to_account_info(),
+            ctx.accounts.system_program.to_account_info(),
+        ];
+        
+        // add "accounts" (hashes) that make up the merkle proof
+        for acc in ctx.remaining_accounts.iter() {
+            accounts.push(AccountMeta::new_readonly(acc.key(), false));
+            account_infos.push(acc.to_account_info());
+        }
+
+        msg!("manual cpi call");
+        solana_program::program::invoke_signed(
+        & solana_program::instruction::Instruction {
+            program_id: ctx.accounts.bubblegum_program.key(),
+            accounts: accounts,
+            data: data,
+        },
+        &account_infos[..],
+        &[&[b"cNFT-vault", &[*ctx.bumps.get("leaf_owner").unwrap()]]])
+        .map_err(Into::into)
+
+        
+    }
+
+    
+    pub fn withdraw_two_cnfts<'info>(ctx: Context<'_, '_, '_, 'info, WithdrawTwo<'info>>,
+        root1: [u8; 32],
+        data_hash1: [u8; 32],
+        creator_hash1: [u8; 32],
+        nonce1: u64,
+        index1: u32,
+        proof_1_length: u8,
+        root2: [u8; 32],
+        data_hash2: [u8; 32],
+        creator_hash2: [u8; 32],
+        nonce2: u64,
+        index2: u32,
+        _proof_2_length: u8 // we don't actually need this (proof_2_length = remaining_accounts_len - proof_1_length)
```

Comment:
Curious why this parameter is included then?

Reply:
because it makes it easier for me to call it ;)
since both trees provide the same data... and if you want to add another one (a third cNFT to send) it is clear what you need..